### PR TITLE
Add flag to allow untrusted ssl connections to build cache

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

File: subprojects/build-cache-http/src/main/java/org/gradle/caching/http/HttpBuildCache.java [hunk 1]
```diff
@@ -85,4 +86,42 @@ public HttpBuildCacheCredentials getCredentials() {
...
+    /**
+     * Specifies whether it is acceptable to communicate with an HTTP build cache backend with an untrusted SSL certificate.
+     * <p>
+     * The SSL certificate for the HTTP build cache backend may be untrusted since it is internally provisioned or a self-signed certificate.
+     * <p>
+     * In such a scenario, you can either configure the build JVM environment to trust the certificate,
+     * or set this property to {@code true} to disable verification of the server's identity.
+     * <p>
+     * Allowing communication with untrusted servers keeps data encrypted during transmission,
+     * but makes it easy for a man-in-the-middle to impersonate the intended server and capture data.
```
Comment (lptr) [inline subprojects/build-cache-http/src/main/java/org/gradle/caching/http/HttpBuildCache.java]: Nit-picking, but "easier" would be more precise.

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java [hunk 2]
```diff
@@ -16,24 +16,37 @@
...
 public class DefaultHttpSettings implements HttpSettings {
     private final Collection<Authentication> authenticationSettings;
     private final SslContextFactory sslContextFactory;
+    private final HostnameVerifier hostnameVerifier;
     private HttpProxySettings proxySettings;
     private HttpProxySettings secureProxySettings;
     private HttpTimeoutSettings timeoutSettings;
 
+    public static DefaultHttpSettings allowUntrustedSslConnections(Collection<Authentication> authenticationSettings) {
+        return new DefaultHttpSettings(authenticationSettings, new AllTrustingSslContextFactory(), ALL_TRUSTING_HOSTNAME_VERIFIER);
```
Comment (lptr) [inline subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java]: Can `AllTrustingSslContextFactory` be a similar constant?

File: subprojects/build-cache-http/src/main/java/org/gradle/caching/http/HttpBuildCache.java (hunk 1 above)
Comment (wolfs) [reply subprojects/build-cache-http/src/main/java/org/gradle/caching/http/HttpBuildCache.java]: fixed.

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java (hunk 2 above)
Comment (wolfs) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java]: I removed the class and made it into a constant.

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java [hunk 3]
```diff
@@ -94,4 +103,44 @@ public boolean verify(String hostname, SSLSession session) {
...
+            @Override
+            public SSLContext get() {
+                try {
+                    SSLContext sslcontext = SSLContext.getInstance("TLS");
+                    sslcontext.init(null, ALL_TRUSTING_TRUST_MANAGER, null);
+                    return sslcontext;
+                } catch (NoSuchAlgorithmException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                } catch (KeyManagementException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
```
Comment (lptr) [inline subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java]: You can simply use `throw new UncheckedException(e)` in these cases. Actually, `throwAsUncheckedException()` is not very useful when you know the type of exception.

Comment (wolfs) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/DefaultHttpSettings.java]: Done that.

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceIntegrationTest.groovy [hunk 4]
```diff
@@ -218,6 +219,41 @@ class HttpBuildCacheServiceIntegrationTest extends AbstractIntegrationSpec imple
...
+        keyStore.enableSslWithServerCert(httpBuildCacheServer)
+        settingsFile.text = useHttpBuildCache(httpBuildCacheServer.uri)
+
+        when:
+        executer.withStackTraceChecksDisabled()
+        withBuildCache().succeeds "jar"
+
+        then:
+        skippedTasks.empty
+        output.contains('sun.security.validator.ValidatorException: PKIX path building failed')
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceIntegrationTest.groovy]: I am checking for `PKIX path building failed: `, which seems to work on IBM JDKs, too: https://builds.gradle.org/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=7849729&_focus=2448

**Pull request merged**